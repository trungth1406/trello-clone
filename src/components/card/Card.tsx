import CardItemList from "./card-item/CardItemList";

export default function Card({card}) {
    return (<section className='card-container'>
        <h3 className='card-header'>Card Title</h3>
        <CardItemList></CardItemList>
    </section>)
}
