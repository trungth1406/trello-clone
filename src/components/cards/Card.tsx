import TaskList from "./tasks/TaskList";
import {useEffect, useState} from "react";

export default function Card({card, onCardDelete}) {

    const [isEditing, setIsEditing] = useState(card.name === null);
    const [cardDetail, setCardDetails] = useState({
        name: card.name,
        items: card.items
    });


    useEffect(() => {
        setIsEditing(cardDetail.name === null);
    }, [cardDetail]);


    const updateCardTitle = (event) => {
        if (event.key === 'Enter') {
            setCardDetails({
                ...cardDetail,
                name: event.target.value
            });
        } else {
            card.name = event.target.value;
        }
    }
    const deleteCard = () => {
        onCardDelete(card);
    }



    return (<li className='card-container' data-testid='card'>
        {isEditing ? <input
                data-testid='card_title_input'
                type="text"
                            className='card-title'
                            placeholder='Card Title'
                            onKeyDown={updateCardTitle}/> :
            <h3 className='card-header' onClick={() => setIsEditing(prev => !prev)}>{cardDetail.name}</h3>
        }
        {cardDetail.name && <TaskList cardDetail={cardDetail}/>}
        <section className='delete-card-button' onClick={deleteCard}>
            <button id='delete_btn'>Delete</button>
        </section>
    </li>)
}
