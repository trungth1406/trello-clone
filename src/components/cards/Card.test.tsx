import { fireEvent, render, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';
import { act } from 'react-dom/test-utils';
import Card from './Card';

// Test the Card component
describe('When Card component mounted', () => {
    render(
        <Card
            card={{ name: null, items: [] }}
            onCardDelete={undefined}
        />
    );

    it('Should show an input field to add a card title when the card name is null', async () => {
        const snapshotComp = renderer.create(
            <Card
                card={{ name: null, items: [] }}
                onCardDelete={undefined}
            />
        );

        expect(screen.getByTestId('card_title_input')).toBeTruthy();
        expect(screen.getByTestId('card_title_input')).toBeInTheDocument();

        expect(screen.getByPlaceholderText('Card Title')).toBeTruthy();

        expect(snapshotComp.toJSON()).toMatchSnapshot();
    });

    it('Should show the card title when focus on the input field and enter a card title', async () => {
        render(
            <Card
                card={{ name: null, items: [] }}
                onCardDelete={undefined}
            />
        );

        fireEvent.focus(screen.getByTestId('card_title_input'));

        fireEvent.change(screen.getByTestId('card_title_input'), {
            target: { value: 'Card Title' },
        });

        fireEvent.keyDown(screen.getByTestId('card_title_input'), {
            key: 'Enter',
            code: 'Enter',
        });

        expect(screen.getByText('Card Title')).toBeTruthy();
    });

    it('Should show card header if card already had a name', async () => {
        // click on the delete button
        render(
            <Card
                card={{ name: 'Card Title', items: [] }}
                onCardDelete={undefined}
            />
        );

        expect(screen.queryByTestId('card_title_input')).toBeNull();
        // Expect the card header to be shown
        expect(screen.getByTestId('card_header')).toBeTruthy();
        expect(screen.getByText('Card Title')).toBeInTheDocument();
    });

    describe('When click on the Card header', () => {
        render(
            <Card
                card={{ name: 'Card Title', items: [] }}
                onCardDelete={undefined}
            />
        );
        it('Should show the input field to edit the card title when click on the card header', async () => {
            render(
                <Card
                    card={{ name: 'Card Title', items: [] }}
                    onCardDelete={undefined}
                />
            );
            fireEvent.click(screen.getByTestId('card_header'));

            expect(screen.getByTestId('card_title_input')).toBeTruthy();
        });

        it('Should update the card header after editing the card title', async () => {
            render(
                <Card
                    card={{ name: 'Card Title', items: [] }}
                    onCardDelete={undefined}
                />
            );
            fireEvent.click(screen.getByTestId('card_header'));

            fireEvent.change(screen.getByTestId('card_title_input'), {
                target: { value: 'New Card Title' },
            });

            fireEvent.keyDown(screen.getByTestId('card_title_input'), {
                key: 'Enter',
                code: 'Enter',
            });

            expect(screen.getByText('New Card Title')).toBeTruthy();
        });
    });

    
    
});
