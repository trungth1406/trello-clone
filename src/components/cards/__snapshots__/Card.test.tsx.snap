// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`When Card component mounted Should show an input field to add a card title when the card name is null 1`] = `
<li
  className="card-container"
  data-testid="card"
>
  <input
    className="card-title"
    data-testid="card_title_input"
    onKeyDown={[Function]}
    placeholder="Card Title"
    type="text"
  />
  <section
    className="delete-card-button"
    onClick={[Function]}
  >
    <button
      data-testid="delete_btn"
      id="delete_btn"
    >
      Delete
    </button>
  </section>
</li>
`;
